### DISCUSSION TITLE
Generate all C structs as immutable with a few exceptions?

### DISCUSSION THREAD

**Discussion by osimonn** [JuliaInterop/Clang.jl#496] [Q&A]
Our project has _a lot_ of C structs to be translated to Julia structs. I just need a few structs to be mutable. Currently, with the `auto_mutability` flag, the opposite of what I am trying to do can be achieved, i.e. all structs are mutable with some exceptions in the `auto_mutability_ignorelist`. Can the opposite be done? All structs are mutable with some exceptions?

It might be straight forward but I tried a bunch of permutations for the `auto_mutability`, `auto_mutability_ignorelist`, and `auto_mutability_includelist` but I cannot get it to do what I want.

Thanks in advance!

**Comment by Gnimuc** [comment]
> All structs are mutable with some exceptions?

In this case, I'd suggest to use a rewriter:

https://github.com/JuliaInterop/Clang.jl/blob/master/gen/generator.jl#L10-L41

I didn't test the following code, but you can get the idea.

```
       build!(ctx, BUILDSTAGE_NO_PRINTING)

        # custom rewriter
        function rewrite!(dag::ExprDAG)
            replace!(get_nodes(dag)) do node
                if some_condition
                    struct_def = node.exprs[1]
                    mutable, name, members = struct_def.args
                    @show mutable, name, members
                    node.exprs[1].args[1] = true
                end
                return node
            end
        end

        rewrite!(ctx.dag)

        build!(ctx, BUILDSTAGE_PRINTING_ONLY)
```

**Reply by Gnimuc** [reply]
instead of `replace!`, a simple for-loop might be enough to do the job.

**Reply by osimonn** [reply]
Thanks @Gnimuc! I will try it a later date since I am busy with other tasks.

**Reply by osimonn** [reply]
I was able to achieve what I wanted using a modified version of the code you suggested:
```julia
build!(ctx, BUILDSTAGE_NO_PRINTING)

# List of structs to force mutability
force_struct_mutability = [:some_c_struct_t]

# custom rewriter
function rewrite!(dag::ExprDAG)
    for node in get_nodes(dag)
        for expr in get_exprs(node)
            if expr.head == :struct
                for s in force_struct_mutability 
                    if expr.args[2] == s
                        expr.args[1] = true # Force mutability
                    end
                end
            end
        end
    end
end

rewrite!(ctx.dag)

build!(ctx, BUILDSTAGE_PRINTING_ONLY)
```
But the problem is that we end up passing this struct to a C call and it crashes. You mentioned the instability of mutable structs briefly in the toml file options [here](https://github.com/JuliaInterop/Clang.jl/blob/f2f4ae43eed608fda8c58208059e15fc201f5b85/gen/generator.toml#L121). Any ideas on how to mitigate this problem?

A solution I have in mind is to create a mutable and immutable versions of the struct(s) in question. When calling a C function, I can use the immutable version, otherwise and I can use the mutable one. However, this can be painful since I have to copy the data back and forth between two structs constantly.
